.buildkite/docker-build-push.nix

```diff
@@ -0,0 +1,94 @@
+# This script will load nix-built docker images of cardano-node
+# into the Docker daemon (must be running), and then push to the
+# Docker Hub. Credentials for the hub must already be installed with
+# "docker login".
+#
+# There is a little bit of bash logic to replace the default repo and
+# tag from the nix-build (../nix/docker.nix).
+#
+# 1. The repo (default "inputoutput/cardano-node") is changed to match
+#    the logged in Docker user's credentials.
+#
+# 2. The tag (default "VERSION") is changed to reflect the
+#    branch which is being built under this Buildkite pipeline.
+#
+#    - If this is a git tag build (i.e. release) then the docker tag
+#      is left as-is.
+#    - If this is a master branch build, then VERSION is replaced with
+#      the git revision.
+#    - Anything else is not tagged and not pushed.
+#
+# 3. After pushing the image to the repo, the "latest" tag is updated.
+#
+#    - "inputoutput/cardano-node:latest" should point to the most
+#      recent VERSION tag build.
+#    - "inputoutput/cardano-node:dev-master" should
+#      point to the most recent master branch build.
+#
+
+{ nodePackages ?  import ../default.nix {}
+, pkgs ? nodePackages.pkgs
+
+# Build system's Nixpkgs. We use this so that we have the same docker
+# version as the docker daemon.
+, hostPkgs ? import <nixpkgs> {}
+
+# Dockerhub repository for image tagging.
+, dockerHubRepoName ? null
+}:
+
+with hostPkgs;
+with hostPkgs.lib;
+
+let
+  images = map impureCreated [
+    nodePackages.dockerImage
+  ];
+
+  # Override Docker image, setting its creation date to the current time rather than the unix epoch.
+  impureCreated = image: image.overrideAttrs (oldAttrs: { created = "now"; }); #### // { inherit (image) version; };
+
+in
+  writeScript "docker-build-push" (''
+    #!${runtimeShell}
+
+    set -euo pipefail
+
+    export PATH=${lib.makeBinPath [ docker gnused ]}
+
+    ${if dockerHubRepoName == null then ''
+    reponame=cardano-node
+    username="$(docker info | sed '/Username:/!d;s/.* //')"
+    fullrepo="$username/$reponame"
+    '' else ''
+    fullrepo="${dockerHubRepoName}"
+    ''}
+
+  '' + concatMapStringsSep "\n" (image: ''
```

Comment:
For reference, here is the conceptual flow for building with the commit hash, then applying conditional tags to allow commit, branch, or version pinning. 

https://github.com/input-output-hk/cardano-graphql/blob/master/Jenkinsfile#L62